src/witness_generator/affine_expression.rs

```diff
@@ -104,6 +105,30 @@ impl AffineExpression {
             .ok_or_else(|| "Cannot solve affine expression.".to_string().into())
     }
 
+    pub fn generate_bit_constraint_from_mask(
```

Comment:
I'm not sure if this is correct and if we should cover more cases? Like `bitconstraint(X << 3, mask) => bitconstraint(X, mask >> 3)`

Reply:
I'm a bit confused so here's what I have:
- `solve_with_bit_constraint` takes an expression `E` which we know equals 0 and tries to learn constraints about the components of `E`, given some existing binary constraints on individual variables
- from the lookup I have an expression `E` and a mask `M` which applies to `E`
- with this new method, I implemented returning a binary constraint if `E` is a single variable
- what I'm missing is that I do not have an expression which is worth 0 and a bunch of binary constraints on individual variables, I have a binary constraint on an expression and I want to learn constraints on individual variables
- So is it correct to say that I cannot use `solve_with_bit_constraints` as is, but rather it needs to be extended to support the input to alternatively be bitconstrained to a given mask, instead of constrained to 0?

Reply:
Oh. Maybe we do have an expression worth 0.
```
(mask => x) <=>
(!x | mask) <=>
!(x & !mask) <=>
(x & !mask) == 0
```

Reply:
True, this is a bit different. You could introduce a new symbolic variable `M`, then use `let expr = l - affine_expression(M);` as the new affine expression, run `expr.solve_with_bit_constraints()` and use the mask for `M`. It sounds a bit complicated, but it would avoid code duplication. Also it would allow you to also use all the other cases:

If we e.g. have `l = 2 * X + 1`, and `r` is constrained to a 0/1, we could even deduce an actual value for `X`, not just a constraint.

For that, you would also need the known constraints on the LHS in the plookup function call - but we could easily extend that.